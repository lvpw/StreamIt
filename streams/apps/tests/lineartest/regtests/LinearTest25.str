/** 
 * test program for doing replacement of fir filters 
 * by a fft, freq multiply and then ifft.
 **/
void->void pipeline LinearTest25 {
   add FloatSource();
   add LowPassFilter(1, pi/2, 20);
   add FloatSink();
 }

 void->float stateful filter FloatSource {
   float x;
   init {
     x = 0;
   }
   work push 1 {
     push(x);
     x = x + 1;
   }
 }

 float->void filter FloatSink {
   work pop 1 {
     println(pop());
   }
 }

/** 
 * Simple FIR low pass filter with gain=g, wc=cutoffFreq(in radians) and N samples.
 * Eg:
 *                 ^ H(e^jw)
 *                 |
 *          ---------------
 *          |      |      |
 *          |      |      |
 *    <-------------------------> w
 *         -wc            wc
 *
 * This implementation is a FIR filter is a rectangularly windowed sinc function 
 * (eg sin(x)/x), which is the optimal FIR low pass filter in 
 * mean square error terms.
 *
 * Specifically, h[n] has N samples from n=0 to (N-1)
 * such that h[n] = sin(cutoffFreq*pi*(n-N/2))/(pi*(n-N/2)).
 * and the field h holds h[-n].
 */
float->float filter LowPassFilter(float g, float cutoffFreq, int N) {
  float[N] h;

  /* since the impulse response is symmetric, I don't worry about reversing h[n]. */
  init {
    int OFFSET = N/2;
    for (int i=0; i<N; i++) {
      int idx = i + 1;
      // generate real part
      if (idx == OFFSET) 
	/* take care of div by 0 error (lim x->oo of sin(x)/x actually equals 1)*/
	h[i] = g * cutoffFreq / pi; 
      else 
	h[i] = g * sin(cutoffFreq * (idx-OFFSET)) / (pi*(idx-OFFSET));
    }
  }

  /* implement the FIR filtering operation as the convolution sum. */
  work peek N pop 1 push 1 {
    float sum = 0;
    for (int i=0; i<N; i++) { 
      sum += h[i]*peek(i);
    }
    push(sum);
    pop();
  }
}
