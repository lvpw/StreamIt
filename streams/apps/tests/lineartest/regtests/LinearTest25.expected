LowPassFilter_1.matrix
[[-2.150286526614309E-8+0.0i]
 [0.03536776453256607+0.0i]
 [1.3913767027418089E-8+0.0i]
 [-0.04547283798456192+0.0i]
 [-1.265269111705436E-9+0.0i]
 [0.06366197764873505+0.0i]
 [1.3913767027418089E-8+0.0i]
 [-0.10610329359769821+0.0i]
 [-1.3913767027418089E-8+0.0i]
 [0.31830987334251404+0.0i]
 [0.5+0.0i]
 [0.31830987334251404+0.0i]
 [-1.3913767027418089E-8+0.0i]
 [-0.10610329359769821+0.0i]
 [1.3913767027418089E-8+0.0i]
 [0.06366197764873505+0.0i]
 [-1.265269111705436E-9+0.0i]
 [-0.04547283798456192+0.0i]
 [1.3913767027418089E-8+0.0i]
 [0.03536776453256607+0.0i]]
LowPassFilter_1.vector
[[0.0+0.0i]]
